File: python/tests/pulse_generator_unittest.py
# @file python/tests/pulse_generator_unittest.py
# @author Guilherme Inacio Goncalves <ggoncalves@iprj.uer.br>
# @copyright Copyright 2022, Guilherme Inacio Goncalves. All rights reserved.
#
# This file is part of CPS (Calorimetry Pulse Simulator).
#
# CPS is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or any later version.
#
# CPS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# CPS. If not, see <https://www.gnu.org/licenses/>.
import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../build/python'))

import unittest
import pycps as lib

class PulseGenerator(unittest.TestCase):

    def setUp(self):
        pulse_shape_file_path = "tests/fixtures/unipolar-pulse-shape.dat"
        self.pulse_shape = lib.TextFilePulseShape(pulse_shape_file_path)

    def test_construction(self):
        pedestal = 15.3
        deformation_level = 0.2
        noise_mean = 0.01
        noise_std_dev = 0.1

        pulse_generator = lib.PulseGenerator(self.pulse_shape)
        pulse_generator.set_pedestal(pedestal)
        pulse_generator.set_deformation_level(deformation_level)
        pulse_generator.set_noise_params(noise_mean, noise_std_dev)
        pulse_generator.set_amplitude_distribution(lib.PulseGenerator.UNIFORM_INT_DISTRIBUTION, [80, 90])
        pulse_generator.set_phase_distribution(lib.PulseGenerator.NORMAL_DISTRIBUTION, [0, 4])

        self.assertEqual(pulse_generator.get_pedestal(), pedestal)
        self.assertEqual(pulse_generator.get_deformation_level(), deformation_level)
        self.assertEqual(pulse_generator.get_noise_mean(), noise_mean)
        self.assertEqual(pulse_generator.get_noise_std_dev(), noise_std_dev)
        self.assertEqual(pulse_generator.get_amplitude_distribution(), lib.PulseGenerator.UNIFORM_INT_DISTRIBUTION)
        self.assertEqual(pulse_generator.get_phase_distribution(), lib.PulseGenerator.NORMAL_DISTRIBUTION)

    def test_generate_pulse(self):
        pulse_generator = lib.PulseGenerator(self.pulse_shape)
        pulse_generator.set_amplitude_distribution(lib.PulseGenerator.UNIFORM_INT_DISTRIBUTION, [10, 20])
        pulse_generator.set_phase_distribution(lib.PulseGenerator.UNIFORM_INT_DISTRIBUTION, [-4, 4])

        # deterministic values
        deterministic_amplitude = 500
        deterministic_phase = 1.0
        analog_pulse = pulse_generator.generate_pulse(deterministic_amplitude, deterministic_phase)
        self.assertEqual(analog_pulse.get_sample(-1), deterministic_amplitude)

        # random values
        analog_pulse = pulse_generator.generate_pulse()
        self.assertGreaterEqual(analog_pulse.get_amplitude(), 10)
        self.assertLessEqual(analog_pulse.get_amplitude(), 20)
        self.assertGreaterEqual(analog_pulse.get_phase(), -4)
        self.assertLessEqual(analog_pulse.get_phase(), 4)



if __name__ == '__main__':
    unittest.main()

File: tests/pulse_generator_unittest.cpp
#include <gtest/gtest.h>
#include "TextFilePulseShape.h"
#include "PulseGenerator.h"

#ifndef PULSE_SHAPE_FILE_PATH
#define PULSE_SHAPE_FILE_PATH "tests/fixtures/unipolar-pulse-shape.dat"
#endif

using namespace cps;

const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);

TEST(PulseGenerator, Pedestal) {
  PulseGenerator pulseGenerator(pulseShape);
  // default value
  EXPECT_EQ(pulseGenerator.GetPedestal(), 0.0);
  // new value
  double pedestal = 50;
  EXPECT_EQ(pulseGenerator.SetPedestal(pedestal).GetPedestal(), pedestal);
}

TEST(PulseGenerator, DeformationLevel) {
  PulseGenerator pulseGenerator(pulseShape);
  // default value
  EXPECT_EQ(pulseGenerator.GetDeformationLevel(), 0.0);
  // new value
  double deformationLevel = 0.1;
  EXPECT_EQ(pulseGenerator.SetDeformationLevel(deformationLevel).GetDeformationLevel(), deformationLevel);
}

TEST(PulseGenerator, NoiseParams) {
  PulseGenerator pulseGenerator(pulseShape);
  // default value
  EXPECT_EQ(pulseGenerator.GetNoiseMean(), 0);
  EXPECT_EQ(pulseGenerator.GetNoiseStdDev(), 0);
  // new value
  double noiseMean = 1.0;
  double noiseStdDev = 3.5;
  pulseGenerator.SetNoiseParams(noiseMean, noiseStdDev);
  EXPECT_EQ(pulseGenerator.GetNoiseMean(), noiseMean);
  EXPECT_EQ(pulseGenerator.GetNoiseStdDev(), noiseStdDev);
}

TEST(PulseGenerator, AmplitudeDistribution) {
  PulseGenerator pulseGenerator(pulseShape);
  // default value
  EXPECT_EQ(pulseGenerator.GetAmplitudeDistribution(), PulseGenerator::UNIFORM_REAL_DISTRIBUTION);
  EXPECT_EQ(pulseGenerator.GetAmplitudeDistributionParams(), std::vector<double>({0, 1023}));
  // new value
  auto distribution = PulseGenerator::UNIFORM_INT_DISTRIBUTION;
  auto params = std::vector<double>{80, 90};
  pulseGenerator.SetAmplitudeDistribution(distribution, params);
  EXPECT_EQ(pulseGenerator.GetAmplitudeDistribution(), distribution);
  EXPECT_EQ(pulseGenerator.GetAmplitudeDistributionParams(), params);
}

TEST(PulseGenerator, PhaseDistribution) {
  PulseGenerator pulseGenerator(pulseShape);
  // default value
  EXPECT_EQ(pulseGenerator.GetPhaseDistribution(), PulseGenerator::NO_DISTRIBUTION);
  // new value
  auto distribution = PulseGenerator::UNIFORM_INT_DISTRIBUTION;
  auto params = std::vector<double>{-4, 4};
  pulseGenerator.SetPhaseDistribution(distribution, params);
  EXPECT_EQ(pulseGenerator.GetPhaseDistribution(), distribution);
  EXPECT_EQ(pulseGenerator.GetPhaseDistributionParams(), params);
}

TEST(PulseGenerator, GeneratePulse) {
  PulseGenerator pulseGenerator = PulseGenerator(pulseShape)
    .SetAmplitudeDistribution(PulseGenerator::UNIFORM_INT_DISTRIBUTION, {80, 90})
    .SetPhaseDistribution(PulseGenerator::UNIFORM_INT_DISTRIBUTION, {-4, 4});

  AnalogPulse* analogPulse;

  // deterministic values
  double deterministicAmplitude = 500;
  double deterministicPhase = 1.0;
  analogPulse = pulseGenerator.GeneratePulse(500, 1.0);
  EXPECT_EQ(analogPulse->GetSample(-1.0), deterministicAmplitude);

  // random values
  analogPulse = pulseGenerator.GeneratePulse();
  EXPECT_GE(analogPulse->GetAmplitude(), 80);
  EXPECT_LE(analogPulse->GetAmplitude(), 90);
  EXPECT_GE(analogPulse->GetPhase(), -4);
  EXPECT_LE(analogPulse->GetPhase(), 4);
}
